x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobConfigProvider.java

```diff
@@ -0,0 +1,396 @@
+                            .setVersion(version)
+                            .request();
+
+                    executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest, ActionListener.wrap(
+                            indexResponse -> {
+                                assert indexResponse.getResult() == DocWriteResponse.Result.UPDATED;
+                                updatedJobListener.onResponse(updatedJob);
+                            },
+                            updatedJobListener::onFailure
+                    ));
+
+                } catch (IOException e) {
+                    updatedJobListener.onFailure(
+                            new ElasticsearchParseException("Failed to serialise job with id [" + jobId + "]", e));
+                }
+
+
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                updatedJobListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Expands an expression into the set of matching names. {@code expresssion}
+     * may be a wildcard, a job group, a job ID or a list of those.
+     * If {@code expression} == ALL or * then all job IDs are returned.
+     * Job groups are expanded to all the jobs IDs in that group.
+     *
+     * For example, given a set of names ["foo-1", "foo-2", "bar-1", bar-2"],
+     * expressions resolve follows:
+     * <ul>
+     *     <li>"foo-1" : ["foo-1"]</li>
+     *     <li>"bar-1" : ["bar-1"]</li>
+     *     <li>"foo-1,foo-2" : ["foo-1", "foo-2"]</li>
+     *     <li>"foo-*" : ["foo-1", "foo-2"]</li>
+     *     <li>"*-1" : ["bar-1", "foo-1"]</li>
+     *     <li>"*" : ["bar-1", "bar-2", "foo-1", "foo-2"]</li>
+     *     <li>"_all" : ["bar-1", "bar-2", "foo-1", "foo-2"]</li>
+     * </ul>
+     *
+     * @param expression the expression to resolve
+     * @param allowNoJobs if {@code false}, an error is thrown when no name matches the {@code expression}.
+     *                     This only applies to wild card expressions, if {@code expression} is not a
+     *                     wildcard then setting this true will not suppress the exception
+     * @param listener The expanded job IDs listener
+     * @return the set of matching names
+     */
+    public void expandJobsIds(String expression, boolean allowNoJobs, ActionListener<Set<String>> listener) {
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().query(buildQuery(expression));
+        sourceBuilder.sort(Job.ID.getPreferredName());
+        String [] includes = new String[] {Job.ID.getPreferredName()};
+        sourceBuilder.fetchSource(includes, null);
+
+        SearchRequest searchRequest = client.prepareSearch(AnomalyDetectorsIndex.configIndexName())
+                .setIndicesOptions(IndicesOptions.lenientExpandOpen())
+                .setSource(sourceBuilder).request();
+
+        final boolean isWildCardExpression = isWildCard(expression);
```

Comment:
I think we can replace this by simply looking whether there is a `*` in the expression and then the behaviour will remain the same. We could even reuse `Regex.isSimpleMatchPattern` to the whole expression untokenized.

Reply:
Actually that's not correct. We need to check for each token in the expression so scrap that.